core/src/main/scala/filodb/core/utils/ColumnFilterMap.scala

```diff
@@ -0,0 +1,197 @@
+ *
+ * Convenient for use-cases that require e.g. configuration of rules against
+ *   label-value filters (filter->rule pairs). The {@link #get} method
+ *   accepts a set of label-value pairs, identifies any matching filter sets,
+ *   then returns the corresponding element.
+ *
+ * For example, consider the (filters,element) pairs:
+ *     - ({l1="v1",l2=~"v2.*"}, elt1)
+ *     - ({l1="v1",l2="v3"},    elt2)
+ *     - ({l2=~"v4.*"},         elt3)
+ *
+ * The following {label:value} sets would map to these elements:
+ *     - {l1=v1,l2=v2,l3=v3} -> elt1
+ *     - {l1=v1}             -> (None)
+ *     - {l2=v4A}            -> elt3
+ *
+ * A label-value map "matches" a set of filters when:
+ *     - all filter labels are present in the map
+ *     - every filter is matched by label in the map.
+ *
+ * NOTE: at most one ColumnFilter in a set can be non-Equals.
+ *
+ * NOTE: This implementation is intended to favor simplicity over efficiency.
+ *   Other implementations should be considered if the {@link #get} runtime
+ *   (described in the method's javadoc) is insufficient.
+ *
+ * @tparam T the type of elements to store/return.
+ * @param filterElementPairs (filter-set,element) pairs; at most one of filter-set
+ *                           can be a non-Equals filter.
+ */
+class ColumnFilterMap[T](filterElementPairs: Iterable[(Iterable[ColumnFilter], T)]){
+
+  // The main idea:
+  // - Require that filter sets include at most one non-Equals filter.
+  // - For each set of filters, deterministically extract a "key" sequence of label names
+  //   and map each key to an inner hash-map. Label names are sorted alphabetically, then any
+  //   non-Equals filters are moved to the end. For example, the following filter sets
+  //   would have the keys:
+  //     1) {labelA="value1a", labelB="value2a"} -> (labelA, labelB)
+  //     2) {labelA="value1b", labelB="value2b"} -> (labelA, labelB)  // Same as above.
+  //     3) {labelC=~"value3", labelD="value4", labelE="value5"} -> (labelD, labelE, labelC)  // Regex at end.
+  //     4) {labelC=~"value3", labelD=~"value4", labelE="value5"} -> INVALID! // >1 regex filter.
+  // - The inner hash-maps are keyed on a prefix of these values (all except the last), and
+  //   the final filter-- along with the element to store-- is stored as the value. For example,
+  //   the inner hash-map of example (3) above would be keyed on (labelD-value, labelE-value) and
+  //   store (labelC-filter, element) pairs.
+  // - When a set of labels-value pairs is processed through the ColumnFilterMap, the values are
+  //   used to access each inner map until (a) the inner map contains a corresponding value, and
+  //   (b) the final filter matches the final label-value. When a match is found, the filter's paired
+  //   element is returned.
+  //
+  // As an example, consider a ColumnFilterMap composed of these entries:
+  //
+  //     {l1="v1", l2="v2", l3="v3a"} -> A
```

Comment:
noob question: what are the `element` that we are referencing  here - A, B, C etc ? ( I am confused since the data structure already stores label name and label values )

Reply:
Seeing from below, `ExportTableConfig` and `Seq(TargetSchemaChange)` looks like the element stored essentially ?

Reply:
That's right -- those two will be the most-commonly stored values.